allennlp/nn/decoding/maximum_marginal_likelihood.py

```diff
@@ -33,43 +34,50 @@ def decode(self,
                decode_step: DecoderStep,
                targets: torch.Tensor = None,
                target_mask: torch.Tensor = None) -> Dict[str, torch.Tensor]:
-        if targets.dim() != 2:
-            raise NotImplementedError("This implementation cannot yet handle batched inputs")
         allowed_transitions = self._create_allowed_transitions(targets, target_mask)
         finished_states = []
         states = [initial_state]
         step_num = 0
         while states:
             step_num += 1
             next_states = []
-            for state in states:
-                allowed_actions = allowed_transitions[tuple(state.action_history)]
-                generator = decode_step.take_step(state, allowed_actions)
-                while True:
-                    try:
-                        next_state = next(generator)
-                    except StopIteration:
-                        break
-                    if next_state.is_finished():
-                        finished_states.append(next_state)
-                    else:
-                        next_states.append(next_state)
+            # We group together all current states to get more efficient (batched) computation.
+            grouped_state = states[0].combine_states(states)
+            allowed_actions = self._get_allowed_actions(grouped_state, allowed_transitions)
+            generator = decode_step.take_step(grouped_state, allowed_actions)
+            while True:
```

Comment:
A for loop over the generator seems a little more pythonic here.

Reply:
That was what I did at first, but something about the way I did it was wrong, and it didn't work.  For a different reason, I removed the generator entirely and now just return a list, so I do the right thing when looping over it.